libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp

```diff
@@ -48,36 +48,67 @@ bool AP_RangeFinder_LeddarOne::detect(RangeFinder &_ranger, uint8_t instance, AP
 // read - return last value measured by sensor
 bool AP_RangeFinder_LeddarOne::get_reading(uint16_t &reading_cm)
 {
+    uint8_t number_detections;
+    LeddarOne_Status leddarone_status;
+
     if (uart == nullptr) {
         return false;
     }
 
-    // send a request message for Modbus function 4
-    if (send_request() != LEDDARONE_OK) {
-        // TODO: handle LEDDARONE_ERR_SERIAL_PORT
-        return false;
-    }
+    switch (modbus_status) {
 
-    uint32_t start_ms = AP_HAL::millis();
-    while (!uart->available()) {
-        // wait up to 200ms
-        if (AP_HAL::millis() - start_ms > 200) {
-            return false;
-        }
-    }
+        case LEDDARONE_MODBUS_PRE_SEND_REQUEST:
+            // clear buffer and buffer_len
+            memset(read_buffer, 0, sizeof(read_buffer));
+            read_len = 0;
 
-    // parse a response message, set number_detections, detections and sum_distance
-    // must be signed to handle errors
-    uint8_t number_detections;
-    if (parse_response(number_detections) != LEDDARONE_OK) {
-        // TODO: when (not LEDDARONE_OK) handle LEDDARONE_ERR_
-        return false;
-    }
+            // send a request message for Modbus function 4
+            if (send_request() != LEDDARONE_OK) {
+                // TODO: handle LEDDARONE_ERR_SERIAL_PORT
+                break;
+            }
+            modbus_status = LEDDARONE_MODBUS_SENT_REQUEST;
+            last_sending_request_ms = AP_HAL::millis();
+            break;
+
+        case LEDDARONE_MODBUS_SENT_REQUEST:
+            if (uart->available()) {
+                // change mod_bus status to read available buffer
+                modbus_status = LEDDARONE_MODBUS_AVAILABLE;
+            } else {
+                if (AP_HAL::millis() - last_sending_request_ms > 200) {
+                    // reset mod_bus status to read new buffer
+                    modbus_status = LEDDARONE_MODBUS_PRE_SEND_REQUEST;
+                }
+            }
+            break;
 
-    // calculate average distance
-    reading_cm = sum_distance / number_detections;
+        case LEDDARONE_MODBUS_AVAILABLE:
+            // parse a response message, set number_detections, detections and sum_distance
+            leddarone_status = parse_response(number_detections);
 
-    return true;
+            if (leddarone_status == LEDDARONE_OK) {
+                reading_cm = sum_distance / number_detections;
+
+                // reset mod_bus status to read new buffer
+                modbus_status = LEDDARONE_MODBUS_PRE_SEND_REQUEST;
+
+                return true;
+            }
+            // keep reading next buffer
+            else if (leddarone_status == LEDDARONE_READING_BUFFER) {
```

Comment:
wouldn't this be simplified by using != READING_BUFFER and moving the logic up into this line? An else break is more logical